### PR TITLE
UCT/CM/RDMACM: share dummy CQ per device

### PR DESCRIPTION
**Pull request by evgeny-leksikov**
## What
share dummy CQ per device

## Why ?
to reduce resources consumption

## How ?
ported from https://github.com/yosefe/ucx/pull/105

### CODE REVIEW

**Comment by yosefe** [comment]
@evgeny-leksikov was it clean CP?

**Comment by evgeny-leksikov** [comment]
> @evgeny-leksikov was it clean CP?

no, there were 2 conflicts around dev_name parameter

**Comment by yosefe** [comment]
bot:pipe:retest

File: src/uct/ib/rdmacm/rdmacm_cm.c [hunk 1]
```diff
@@ -624,6 +674,8 @@ UCS_CLASS_CLEANUP_FUNC(uct_rdmacm_cm_t)
 
     ucs_trace("destroying event_channel %p on cm %p", self->ev_ch, self);
     rdma_destroy_event_channel(self->ev_ch);
+
```
Comment (yosefe) [inline src/uct/ib/rdmacm/rdmacm_cm.c:677]: remove space line

File: src/uct/ib/rdmacm/rdmacm_cm.h [hunk 2]
```diff
@@ -7,16 +7,23 @@
 #ifndef UCT_RDMACM_CM_H
 #define UCT_RDMACM_CM_H
 
-#include <uct/base/uct_cm.h>
 #include "rdmacm_def.h"
 
+#include <uct/base/uct_cm.h>
+#include <ucs/datastruct/khash.h>
+
+#include <infiniband/verbs.h>
```
Comment (yosefe) [inline src/uct/ib/rdmacm/rdmacm_cm.h:15]: IMO it's not needed

File: src/uct/ib/rdmacm/rdmacm_cm.c [hunk 3]
```diff
@@ -58,6 +58,53 @@ ucs_status_t uct_rdmacm_cm_reject(struct rdma_cm_id *id)
...
 }
 
+ucs_status_t uct_rdmacm_cm_get_cq(uct_rdmacm_cm_t *cm, struct ibv_context *verbs,
+                                  const char *device_name, struct ibv_cq **cq_p)
+{
+    struct ibv_cq *cq;
+    khiter_t iter;
+    int ret;
+
+    iter = kh_put(uct_rdmacm_cm_cqs, &cm->cqs, device_name, &ret);
```
Comment (yosefe) [inline src/uct/ib/rdmacm/rdmacm_cm.c:68]: how does it work? looks like the hash table key can point to memory on the stack (device_name->dev_name)
IMO better use rdma_cm_id::pd pointer as the key
Any change it's using stack garbage as the key and actually creating new CQ every time?

**Comment by evgeny-leksikov** [comment]
re-cherry-picked squashed https://github.com/yosefe/ucx/pull/105 + https://github.com/yosefe/ucx/pull/113 to avoid conflicts which were appeared due to wrong approach with `dev_name`

File: src/uct/ib/rdmacm/rdmacm_cm_ep.c [hunk 4]
```diff
@@ -199,41 +180,36 @@ static ucs_status_t uct_rdmacm_cm_create_dummy_cq_qp(struct rdma_cm_id *id,
...
+
+    status = uct_rdmacm_cm_get_cq(uct_rdmacm_cm_ep_get_cm(cep), cep->id->verbs,
+                                  cep->id->pd->handle, &cq);
+    if (status != UCS_OK) {
+        return status;
+    }
 
     /* create a dummy qp in order to get a unique qp_num to provide to librdmacm */
-    status = uct_rdmacm_cm_create_dummy_cq_qp(cep->id, &cq, &qp);
+    status = uct_rdmacm_cm_create_dummy_qp(cep->id, cq, &cep->qp);
```
Comment (yosefe) [inline src/uct/ib/rdmacm/rdmacm_cm_ep.c:207]: maybe call uct_rdmacm_cm_get_cq from uct_rdmacm_cm_create_dummy_qp()?

Comment (evgeny-leksikov) [reply src/uct/ib/rdmacm/rdmacm_cm_ep.c:207]: I would like to avoid hidden side effect of such function, let's keep things simpler

Comment (yosefe) [reply src/uct/ib/rdmacm/rdmacm_cm_ep.c:207]: ok

**Review by yosefe — APPROVED** [review]

**Pull request merged**